docs/codeql/codeql-language-guides/customizing-library-models-for-java.rst

```diff
@@ -0,0 +1,403 @@
+.. _customizing-library-models-for-java:
+
+Customizing Library Models for Java
+===================================
+
+.. include:: ../reusables/beta-note-customizing-library-models.rst
+
+The Java analysis can be customized by adding library models (summaries, sinks and sources) in data extension files.
+
+A data extension file for Java is a YAML file in the form:
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/java-all
+         extensible: <name of extension point>
+       data:
+         - <tuple1>
+         - <tuple2>
+         - ...
+
+The data extension can contribute to the following extension points:
+
+- **sourceModel**\(package, type, subtypes, name, signature, ext, output, kind, provenance)
+- **sinkModel**\(package, type, subtypes, name, signature, ext, input, kind, provenance)
+- **summaryModel**\(package, type, subtypes, name, signature, ext, input, output, kind, provenance)
+- **neutralModel**\(package, type, name, signature, provenance)
+
+TODO: Link or inline documentation on how to add dataextensions.
+Are we going for extensions packs as the recommended default?
+If yes, then we probably need to elaborate with a concrete example.
```

Comment:
In order to avoid duplication, we should probably be including links, though a one sentence description of what an extensible predicate is would be helpful.

Reply:
I will add a link to a reusable (like the beta notice).

Reply:
To avoid making this an eternity PR, should we just add a quote in the reusable stating that the data extensions is documented in a google doc for now?

Reply:
> To avoid making this an eternity PR, should we just add a quote in the reusable stating that the data extensions is documented in a google doc for now?

Yes!